### ISSUE TITLE
feature request: frequency response measurer

### ISSUE THREAD

**Issue by squinkylabs** [bogaudio/BogaudioModules#116]
To measure the freq response of a filter you can run white noise into it, then into your FFT analyzer. But it's sub-optimal for a few reasons:
* The output is noisy, so it's hard to make out the details.
* The extra noise masks any noise your filter might make.
* White noise it a pretty bad tests signal.

If you actually wanted to do this (a big if, I admit) "all" you would have to do is:
* Output a stimulus signal (exponentially swept sine is the best).
* Before you display the FFT of then input, divide it by the FFT of the stimulus.

If you are interested it is super easy. I've got a pretty readable implementation in my repo, so if it would save you any time/effort I could point that out.

**Comment by MarcBoule** [comment]
I could definitely use this also! :-)

**Comment by mdemanett** [comment]
I'm not sure I understand what this would do / how it would work.  Is there something you could point me to to read up on it?  Also yeah I'd take a look at your implementation.

Without understanding it, maybe something like this would work:
  - A new separate module to generate the signal (I was watching some Hainbach stuff and had a vague idea about making some "test equipment" modules...though test equipment seems to mainly be just really accurate oscillators and filters...anyway a swept sine sounds like a candidate for something like that).
  - Add an option to the analyzers to divide input A by input B, and so on.

**Comment by squinkylabs** [comment]
Interesting. I made a separate module to generate a swept sine and tried it, but since it wasn't flat the results were bad. Also it's best if the source knows how big the FFT frames are - it only works because the stimulus and the FFT are synchronous. That way you don't need to worry about windowing.

My current one is just used for unit tests. I take the freq response of the code I'm testing and see if it's correct. It's all in a class called Analyzer in the test folder. Here's the (simple) source for doing the freq response:  https://github.com/squinkylabs/SquinkyVCV/blob/master/test/Analyzer.cpp#L243

I'll see if I can find some background info....

**Comment by squinkylabs** [comment]
When you do this (correctly) the resulting fft magnitude is very smooth - it looks like a textbook. I think for you do to it you would want to put the generator of the stimulus function in your module, so you would know they are the same size. I don't know if they need to start recording on the same sample - I suspect a constant delay cancels out, but I've never investigated that.

You can use any stimulus at all, as long as it generates all the freq you care about. normal white noise will usually have some freq where there is no output, so you get divide by zero errors. I used to use "fake" white noise (like I use in my Colors module). Then I used fake pink noise. Then a co-worker showed me that swept sin is better, so now I use that.

**Comment by mdemanett** [comment]
I finally had some time to look at this.  Super interesting.  This seems to lay out it out in detail:
[https://hal.archives-ouvertes.fr/hal-02504321/document](https://hal.archives-ouvertes.fr/hal-02504321/document)

@squinkylabs a question, in your code looks like you use 16*1024 samples for the FFT buffer and sweep 20-20k hz over that.  Any significance to that size other than it works?  (Obvs it has to be a power of two.)

**Comment by squinkylabs** [comment]
Oh, nice article! As to your question - no significance. I wanted it nice and big so I could get detail, and it doesn't matter much how long it takes to run (in that case). And, if you use my method of dividing the FFT by the FFT of the sweep, almost nothing matters. Of course you want it to have significant energy in all the frequencies you care about, so I go from 20 to 20. And I do exponential sweep so that I have roughly  a "pink" distribution (equal energy per octave).

It's true that if you pick the sweep just right then some of the math can work out a lot easier and more efficient, but the brute force way I do it works for me, so I've never tried to get tricky.

I don't think my dumb little sweep generator cares about powers of two or anything. At least I don't think it does.

**Comment by mdemanett** [comment]
I finally got around to this:

![Screen Shot 2020-08-19 at 8 01 24 AM](https://user-images.githubusercontent.com/49971/90632336-3bcf5480-e1f2-11ea-9a0c-d55093a2c088.png)

There's a test build [here](https://github.com/bogaudio/BogaudioModules/releases/tag/v1.1.34-test3), and preliminary docs [here](https://github.com/bogaudio/BogaudioModules/blob/master/README-prerelease.md#-ranalyzer).

Don't know if you guys have moved on, but I'd be curious what you think....

**Comment by MarcBoule** [comment]
Wow, very cool Matt! 

It tried a first test with a unity gain filter (i.e. passthrough!), and I'm not sure I understand what's going on exactly. 
I have Freq1 set at 100Hz and Freq2 at 15kHz, and I'm not seeing that square response where I was expecting it (i.e. from 100Hz to 15kHz)

![image](https://user-images.githubusercontent.com/33604154/90640949-a090ac00-e1fe-11ea-9a21-08d873175392.png)

Also, I'm wondering why the test sweep doesn't appear flat in the spectrum, but I'm sure it's just me not understanding something :-)

Great work on the dark panels and this new module, very cool!

**Comment by mdemanett** [comment]
Oof, there was a pretty blatant bug reading the frequency knobs -- thanks for catching that!

Indeed the exponential sweep amplitude declines like that -- switch it to linear and it comes out flat.  But I think that's right -- with the exponential, the sweep spends relatively more time in the lower frequencies, thus putting more energy into the FFT in the lower bands, thus showing higher amplitude there.  I haven't run across a source that could confirm or refute this, though.

**Comment by MarcBoule** [comment]
Thanks, I figured it might be an energy bin thing :-)

Just one last thing, the Freq2 now looks spot on, but I'm still having an offset on Freq1:

![image](https://user-images.githubusercontent.com/33604154/90702669-cc8d4b00-e259-11ea-9872-a822a5c005d5.png)

**Comment by mdemanett** [comment]
I think what you're seeing there is that the orange line is weird, it just being the ratio of the response to the test.  Turn the green (test signal) line display back on and it should be tracking FREQ1.  (Or close, there seems to be a thing at low frequencies where the test graph comes in above the actual frequency...not sure yet if this is a bug or an artifact of the FFT).

Here's an extreme case:

![Screen Shot 2020-08-20 at 7 32 24 PM](https://user-images.githubusercontent.com/49971/90836011-3a01b000-e31c-11ea-889f-088b75916f40.png)

I may just get rid of the orange line, and have the display always just be the response overlaying the test.  I don't think it adds much in terms of information.  And it's strange to be plotting a number without a unit on what is otherwise a frequency/amplitude graph.

**Comment by squinkylabs** [comment]
I hate to be a dummy, but how do you use this thing? It doesn't do anything for me, altough if I hit the exp button it crashes:

Assertion failed: time >= minTimeSeconds, file src/dsp/oscillator.cpp, line 363

**Comment by squinkylabs** [comment]
Is t v1.1.34-test3 still the right tag to build in from your repo? Builds just fine, but does absolutely nothing. Do you need to press the trigger button? (that crashes, too).
![image](https://user-images.githubusercontent.com/16314838/92297672-bab4d480-eef6-11ea-8a9c-6e6ff0d7442d.png)

**Comment by squinkylabs** [comment]
tried building master. same results. doesn't do anything or asserts immediately.

**Comment by squinkylabs** [comment]
oops - sorry, hit the close button by mistake.

**Comment by squinkylabs** [comment]
OH, ok, user error - sort of. I had left my VCV set at X16 sample rate. So something bad happens there, but that's fine. (but you should fix whatever that bug is).

So, yes, it works and it's super useful. A couple of things could be better. 1) compensate for the exponential sweep so you show a flat frequency response. 2) get rid of the noise that's on top of both traces (you can see it, above).

There's no need to inject noise here, and I have a feeling that the same thing would fix both of these issues:

freq response = FFT(output) / FFT(innput). If you divide out the FFT it will look flatnot matter what. And if you ever want to show the phase you will need to do that anyway.

I assume that your sweep in synchronous with your FFT window? As long as that's the case you don't need to worry about  windowing and such.

Anyway - super cool. It's going to be extremely useful.

**Comment by squinkylabs** [comment]
btw, although I no longer have a GUI freq response measurer, I do have one with no UI that I used for tests. You can see here how easy it is to divide out the stimulus signal: https://github.com/squinkylabs/SquinkyVCV/blob/43bb0455e469dc50ab06274e9badbcfab465b246/test/Analyzer.cpp#L243

**Comment by squinkylabs** [comment]
totally loving this module! Is this a bug, or something that VCV just does to you? If I set it to loop every now and then there is this burst of "jaggies" in the low frequency. This junk isn't in my filter (I dont think). I rarely see this if I trigger manually:
![image](https://user-images.githubusercontent.com/16314838/92311576-07d38d80-ef6d-11ea-952b-4298f6478535.png)

**Comment by squinkylabs** [comment]
I don't think the frequency controls are super useful. They do let you move the sweep noise to some other frequency, But I think (especially if you are using exp sweep) there is plenty of stimulus energy at every freq, there is no need to have two knobs to steer more energy into a narrower range.

I would really like to see an "ultra ultra" resolution for this and for your regular analyzers so that I could see more detail in the low frequencies. It's pretty common if you make a 20hz HPF and use single precision to get noise in the low frequencies (for example) but with this number of FFT bins you can't see any detail "down low".

And this is just my opinion, but I think if you fixed the issues with exp sweep (get rid of the noise and normalize the output so it's still "flat") then you would have no need for the linear sweep.

Hey, while we are on a wish list, I use your analyzer XL all the time, but I always need to switch it to "ultra". If there were a persistent preference to that it would be great. (I think you have a persistent preference for panel theme already?).

Once again, this plugin is really great. don't know how many people will use it, but I will use it all the time.

oh, wait, another feature I've been dying for! could you use log scaling on the vertical and horizontal axis? 2/3 of the display is 1k and above. and if there vertical had a log scale option you could see filter slopes much easier, and in general it would be better.

oh, is this a bug in the old Analyzer XL? The scale doesn't draw for me at the bottom. Is that bottom bin 50?
![image](https://user-images.githubusercontent.com/16314838/92311763-f12e3600-ef6e-11ea-9f9d-2de5c9f28b91.png)

**Comment by mdemanett** [comment]
Hey, thanks for the comments.  Glad it's somewhat useful.

I fixed the high SR crash.  I increased the FFT size to 32K when the SR is >= 96K.  And added an Ultra+ quality to Analyzer-XL, which again goes to 32K buffers if the SR is high, 16K otherwise. The thing where XL doesn't save its settings is fixed on master already (not on the library build).  

I tried the thing with plotting the ratio of out/in before, build from cd440496575393db6852aa693b38e5e4226f260e to see it, or otherwise it's the orange line in some of the screens above.  I removed it on deciding it's probably more confusing than informative.  Also it bugged me that the ratio has no unit, and plotting a randomly-scaled unitless line on an amplitude/frequency graph felt weird.

At one point I worked out how to make the exp amplitudes flat -- but doing so meant very large gains at high frequencies, such that the test signal was like 100V at points.  That seemed undesirable, and I gave up on it.

**Comment by squinkylabs** [comment]
well output / input is exactly the transfer function of a linear system, and it's exactly what you go want to see!

Found an article on that that says "Swept-sine chirps are analyzed using twochannel FFT techniques (Fig. 4) to determine the
frequency response of the device under test (DUT),
( )
( )
( )
X f
Y f
H f DUT = (6)
where Y( f ) is the FFT of the input
channel, and X ( f ) is the FFT of the
reference channel. Division by the
reference channel response (assuming it
is non-zero) cancels out both the
magnitude and phase irregularities
present in the test signal.2
 This
automatically accounts for any intended or
unintended non-flatness in the stimulus
signal, as well as zeros out any group
delay present. The impulse response of
the DUT is then the inverse-FFT of
H ( f ) DUT .
"

Here's the article: https://www.thinksrs.com/downloads/pdfs/applicationnotes/SR1_SweptSine.pdf

Oh, my, I see that's from a company where me genius friend has been working for the last 20 years!

**Comment by squinkylabs** [comment]
Tried the fixes - works great! btw, I wasn't reporting a "bug"with serializing the quality, I was requesting an esoteric feature: when I insert a new instance of analyzer I would like it to default to my preference. Which has always been "ultra", but will not probably but ultra+.

Oh, and not just somewhat useful - super useful! I would think most plugin devs would use it! I've found it's so easy to make dumb mistakes, like having your filter frequencies off by a factor of two pi!

Thanks again.

**Comment by squinkylabs** [comment]
Oh, and about those large gains trying to correct for the exponential falloff, if I could give a few more pleas to do this:
1) an internal gain of 1000 is no big deal with floating point numbers.
2) Either your eye is going to have to do this calculation, or math will have to do it. I think math can do a better job.
3) If you divide the FFT you will also correct from small imperfections in your chirp signal.
4) Everyone divides out the input. It works!
4) I've even done this using white noise as a stimulus source. white noise is a terrible stimulus and actually does causes S/N issues, but the technique does work with white noise - it works with any stimulus.

**Comment by squinkylabs** [comment]
I'm working on some filters now, and i can't tell you how useful this plugin is!

**Comment by squinkylabs** [comment]
I'm sure you know this, but the decibel scale is completely wrong. I would expect if I wired the send and return together it would be a perfectly flat line at 0 dB, but it isn't:
![image](https://user-images.githubusercontent.com/16314838/93026580-d3d71880-f5bb-11ea-8991-636d3440d4e3.png)

**Comment by squinkylabs** [comment]
The exponential sweep really helps with the low end noise. I would use it, but without a flat display it's too difficult to interpret.

**Comment by mdemanett** [comment]
> I'm sure you know this, but the decibel scale is completely wrong. I would expect if I wired the send and return together it would be a perfectly flat line at 0 dB, but it isn't:

It plots the spectrum of the test signal (green), then the return signal (magenta) -- since they're the same, the return draws exactly over the test.  

If there was a plot of the delta gain in db from test to return, that would plot along 0db here.  That sounds useful.

**Comment by mdemanett** [comment]
I added back the "analysis" plot, this time as the difference (response - test), plotted in decibels.  It makes much more sense:

![Screen Shot 2020-09-16 at 11 12 02 PM](https://user-images.githubusercontent.com/49971/93415951-ae276900-f872-11ea-9da1-0936a2f8807c.png)

One may also just see the analysis, context menu "Display traces" -> "Analysis only":

![Screen Shot 2020-09-16 at 11 14 14 PM](https://user-images.githubusercontent.com/49971/93415974-b8e1fe00-f872-11ea-8286-cb613d0b39f2.png)

**Comment by squinkylabs** [comment]
cool. I'll try it out. what is that stuff the looks like bursts of junk?

**Comment by squinkylabs** [comment]
yeah, this new orange line is quite nice! if you can just get rid of the bursts of junk. looking good.

**Comment by squinkylabs** [comment]
could you point me to where to code lives that computes the orange line? Where are the FFT's actually done in this plugin?

**Comment by mdemanett** [comment]
The junk is apparently an artifact of the FFT, or rather applying the FFT to the test and response signals without windowing them, which is to say applying the FFT with the implicit rectangular window.  So I added a window option, where the window is applied to the signals going into the FFT (not to the test signal going out at SEND).  A window is on by default, but can be turned off (or rather, switched to rectangular) on the context menu.  At the end of the day each choice is a choice between different approximations...but at least the not-rectangle choices get rid of the weird bursts.

**Comment by mdemanett** [comment]
> could you point me to where to code lives that computes the orange line? Where are the FFT's actually done in this plugin?

The orange line calculation is in AnalysisBinsReader in src/Ranalyzer.cpp.  The FFT is ultimately FFTReal under lib/ffft, via src/dsp/analyzer.* and src/analyzer_base.*.

**Comment by squinkylabs** [comment]
well, if that makes it work ok...

I can't figure out your code - it's like mine - way too many classes for my tinny brain to figure out.

When I did this 20 years ago I used no windowing at all. I just made my "chirp" be the size of an FFT frame, and then all I did was:

complex response = complex_fft(captured data) / complex_fft(stimulus).

It had no artifacts whatsoever. I could use any stimulus - it didn't have to be a chirp, I could use noise, anything. Never got and artifacts like those bursties, never windowed anything.

But it sound like the bursts are tamed, so that's cool. Perhaps it will be "shapable" now? or close to it? It's a great tool. tx again.
